app/notifiers/notifier/base.rb

```diff
@@ -0,0 +1,79 @@
+# The base class of notifiers. This is meant to be called by the Notifications Framework
+#
+# @api notifications
+class Notifier::Base
+  class ActivityWrapper < SimpleDelegator
+    # Send notifications according to input type and recipient
+    #
+    # @param [Object] recipient The recipient of the notification
+    # @param [Symbol] type The type of notification
+    def notify(recipient, type)
+      super(recipient, type).tap do |notification|
+        case recipient
+        when Course
+          emails = @notifier.send(:notify_course, notification)
+          @pending_emails += emails unless emails.nil?
+        when User
+          email = @notifier.send(:notify_user, notification)
+          @pending_emails += email unless email.nil?
+        end
+      end
+
+      self
+    end
+
+    # Save activity and send out pending emails
+    def save(*)
+      super.tap do |result|
+        @pending_emails.each(&:deliver_later) if result
+      end
+    end
+
+    private
+
+    def initialize(notifier, activity) #:nodoc:
+      super(activity)
+      @notifier = notifier
+      @pending_emails = []
+    end
+  end
+
+  class << self
+    # This is not meant to be called by user directly
```

Comment:
...document the purpose.

document the "why", not the "how". "how" should be evident in your code.